### PR TITLE
libc_preload: check for C library/kernel differences

### PR DESCRIPTION
**Pull request by sporksmith**
This adds an advisory check in our script for generating syscall wrappers that warns when we generate wrappers for syscalls whose man pages note "C library/kernel differences".

Suppressed most of the syscalls that it was warning about; reviewed some, adding those to either a "definitely suppress" list, or an "allow list".

Removing an incorrect wrapper for `sched_getaffinity` uncovered a bug in our `sched_getaffinity` syscall implementation. Fixes the bug.

### CODE REVIEW

File: src/lib/libc_preload/gen_syscall_wrappers_c.py
Comment (stevenengler) [inline src/lib/libc_preload/gen_syscall_wrappers_c.py:80]: > These wrapper functions (including the one for setuid()) employ a signal-based technique to ensure that when one thread changes credentials, all of the other threads in the process also change their credentials.

This one is kind of funny. It looks like the wrapper sends a signal to all threads, and the threads have signal handlers to change their uid.

**Review by stevenengler — APPROVED** [review]

**Comment by sporksmith** [comment]
The benchmark appears to show a regression:

image

Given the recent weirdness I'm taking this with a grain of salt.

I wasn't able to reproduce a regression locally with the tor-minimal test, but I did go through the syscall counts in the shadow log and checked to see if any of the top ones were affected:

```
read:2266687946
epoll_wait:1192719367
write:909325735
epoll_ctl:463482287
getsockopt:243728020
ioctl:240557914
sendto:224556251
recvfrom:174803279
timerfd_settime:131063909
getrandom:98808317
lseek:39697029
close:28374216
** fstat:13887153  
openat:8235718
timerfd_create:8039732
setsockopt:6350037
socket:5853045
connect:5846462
getsockname:5501671
** uname:5301673
fcntl:5290741
accept:5281822
shutdown:4627709
futex:3698499
accept4:3173674
open:3122655  
* brk:780132
shadow_hostname_to_addr_ipv4:132188
```

```
* syscalls that don't have a wrapper
** syscalls  that had a wrapper  removed in this PR
```

I'll see about adding manual wrappers for `fstat` and `uname`...

**Comment by sporksmith** [comment]
Started a new run: https://github.com/shadow/benchmark/actions/runs/5226688803

**Comment by sporksmith** [comment]
Looks OK now: 
image

Will wait for another round of review before merging, but no rush. Have a good weekend!

**Review by stevenengler — APPROVED** [review]

**Pull request merged**
